### PR TITLE
[cloud_firestore, cloud_firestore_web] Fix encoding issues in DocumentSnapshotPlatform and FieldValueFactoryWeb

### LINKED ISSUE #2044: [cloud_firestore] DocumentReference in startAfterDocument snapshot causes crash (invalid encode argument)
@amrfarid140 @ditman I think I have found another one of the `.valueEncode` issues as discussed [here](https://github.com/FirebaseExtended/flutterfire/pull/1994/files#r379145491). See also: #1986.

The error I get is `Invalid argument: Instance of 'DocumentReference'`.

<details>
<summary>Full stack trace</summary>

```
E/flutter (14093): #0      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:392:7)
E/flutter (14093): #1      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:83:13)
E/flutter (14093): #2      StandardMessageCodec.writeValue.<anonymous closure> (package:flutter/src/services/message_codecs.dart:389:9)
E/flutter (14093): #3      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:379:8)
E/flutter (14093): #4      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:387:13)
E/flutter (14093): #5      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:83:13)
E/flutter (14093): #6      StandardMessageCodec.writeValue.<anonymous closure> (package:flutter/src/services/message_codecs.dart:389:9)
E/flutter (14093): #7      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:379:8)
E/flutter (14093): #8      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:387:13)
E/flutter (14093): #9      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:83:13)
E/flutter (14093): #10     StandardMessageCodec.writeValue.<anonymous closure> (package:flutter/src/services/message_codecs.dart:389:9)
E/flutter (14093): #11     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:379:8)
E/flutter (14093): #12     MapView.forEach (dart:collection/maps.dart:337:10)
E/flutter (14093): #13     StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:387:13)
E/flutter (14093): #14     FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:83:13)
E/flutter (14093): #15     StandardMessageCodec.writeValue.<anonymous closure> (package:flutter/src/services/message_codecs.dart:389:9)
E/flutter (14093): #16     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:379:8)
E/flutter (14093): #17     StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:387:13)
E/flutter (14093): #18     FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:83:13)
E/flutter (14093): #19     StandardMethodCodec.encodeMethodCall (package:flutter/src/services/message_codecs.dart:524:18)
E/flutter (14093): #20     MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:148:13)
E/flutter (14093): #21     MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:329:12)
E/flutter (14093): #22     MethodChannel.invokeMapMethod (package:flutter/src/services/platform_channel.dart:356:48)
E/flutter (14093): #23     MethodChannelQuery.getDocuments (package:cloud_firestore_platform_interface/src/method_channel/method_channel_query.dart:87:46)
E/flutter (14093): #24     Query.getDocuments (package:cloud_firestore/src/query.dart:37:34)
```

</details>

### Setup

The error happens every time a document has a `DocumentReference` as a value for a field and the snapshot is passed to `startAfterDocument` with this line (the `getDocuments` call causes it):

```dart
query.startAfterDocument(documentSnapshot).limit(8).getDocuments()
```

This was recently introduced with `0.13.0`. I am currently using `cloud_firestore` version `0.13.3`.

### PR DESCRIPTION
**Pull request by ditman**
## Description

This change reproduces the issue reported in #2044 with an Integ test, and fixes it.

While running Integ tests, I discovered some extra changes required to get them all to pass (on Android, at least), and shaved that Yak.

This also fixes another encoding issue in _web that was recently introduced. Web still has one broken Integ test (re: encoding of Timestamp objects), that I'll attempt to fix next.

/cc @amrfarid140 

## Related Issues

Closes #2044. 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/cloud_firestore/cloud_firestore/lib/src/utils/platform_utils.dart [hunk 1]
```diff
@@ -9,7 +9,10 @@ class _PlatformUtils {
           DocumentSnapshot documentSnapshot) =>
       platform.DocumentSnapshotPlatform(
           documentSnapshot.reference.path,
-          documentSnapshot.data,
+          // We could access `documentSnapshot._delegate.data` directly instead
+          // of going through the getter that `replaceDelegatesWithValuesInMap`
+          // on the data, but this way, the code is not tied to a part-ed lib implementation.
+          _CodecUtility.replaceValueWithDelegatesInMap(documentSnapshot.data),
```
Comment (ditman) [inline packages/cloud_firestore/cloud_firestore/lib/src/utils/platform_utils.dart]: I need to bump the changelog and version of cloud_firestore as well!

**Comment by ditman** [comment]
This doesn't seem like a flake, but I don't know how to fix this :/ (the error seems to come from Firebase core?)

```
** BUILD FAILED **
The following build commands failed:
	CompileC /private/var/folders/3y/l0z1x3693dl_8n0qybp4dqwh0000gn/T/cirrus-ci-build/packages/cloud_firestore/cloud_firestore/example/build/macos/Build/Intermediates.noindex/Pods.build/Release/firebase_core.build/Objects-normal/x86_64/FLTFirebaseCorePlugin.o /Users/anka/.pub-cache/hosted/pub.dartlang.org/firebase_core-0.4.4+2/darwin/Classes/FLTFirebaseCorePlugin.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler
```

**Comment by collinjackson** [comment]
That mac failure seems to be affecting all builds right now. @franciscojma86 any ideas?

**Comment by franciscojma86** [comment]
I just ran it locally and I'm getting `UserAgent.h not found`. https://github.com/FirebaseExtended/flutterfire/pull/2099 should fix it when it lands?

**Comment by ditman** [comment]
@franciscojma86 #2099 is failing with the same error. Is this one of those cases where we need to merge in red to get the tree green again?

**Comment by franciscojma86** [comment]
Not quite sure. I seems like @collinjackson still need to update the PR with Jenn's comments.

**Comment by collinjackson** [comment]
I'll try to get to #2099 today. Thanks for your patience.

**Comment by ditman** [comment]
@collinjackson @franciscojma86 is it OK if I merge this with that check failing? Several other (seemingly unrelated) changes have been landed with that test in red.

**Comment by collinjackson** [comment]
Yes, go ahead and land it, the tests are red on tip of tree so fixing so landing #2099 isn't going to improve the situation, it'll just unblock @franciscojma86 troubleshooting the failing mac test.

I think that what may have happened is that https://github.com/FirebaseExtended/flutterfire/commit/7a312ce22bd5e19dd65b907d5e4d57132675a9d9 passed the tests but then broke the build shortly afterwards when it was published. Having a way of testing the effect of firebase_core changes on other plugins in this repo prior to publication would be helpful here.

We could roll back that PR to do so would effectively mean not having macos support and might do more harm than good.

**Comment by ditman** [comment]
> We could roll back that PR to do so would effectively mean not having macos support and might do more harm than good.

@collinjackson I think we should try to fix ASAP and roll forward, if possible (but that's a discussion for another ticket PR/Issue)

**Comment by ditman** [comment]
I'm landing this with the `build-apps+drive-examples` check failing.

**Comment by ditman** [comment]
I'll wait for the post-submit to finish before publishing this two plugins.

**Comment by ditman** [comment]
The post-submit is [not worse](https://github.com/FirebaseExtended/flutterfire/runs/491015533) than the [previous one](https://github.com/FirebaseExtended/flutterfire/runs/486498699), so I'm publishing this two plugins now.

**Comment by ditman** [comment]
Tagged and published:

* https://pub.dev/packages/cloud_firestore/versions/0.13.4+1
* https://pub.dev/packages/cloud_firestore_web/versions/0.1.1+1

**Pull request merged**